internal/xpkg/fetch.go

```diff
@@ -165,6 +188,79 @@ func (i *K8sFetcher) Fetch(ctx context.Context, ref name.Reference, secrets ...s
 	)
 }
 
+// verifyOCISourceSignature verifies the authenticity of the given image reference url. First, it tries to keyful approach
+// by looking at whether the given secret exists. Then, if it does not exist, it pushes a keyless approach for verification.
+func (i *K8sFetcher) verifyOCISourceSignature(ctx context.Context, keychain authn.Keychain, ref name.Reference, vp string, vs []string) error {
+	// Verify the image
+	if i.validatePackages {
+		switch vp {
+		case "cosign":
+			authnKeychain := soci.WithAuthnKeychain(keychain)
+
+			if len(vs) > 0 {
+				//TODO(JesseSanford): verify the following copypasta is appropriate:
+				// https://github.com/google/go-containerregistry/blob/9e939fbf5b5903da81f258b74ef509488569d411/pkg/authn/kubernetes/keychain.go#L80-L91
+				var validationSecrets []corev1.Secret
+				for _, name := range vs {
+					secret, err := i.client.CoreV1().Secrets(i.namespace).Get(ctx, name, metav1.GetOptions{})
+					if k8serrors.IsNotFound(err) {
+						//logs.Warn.Printf("secret %s/%s not found; ignoring", opt.Namespace, name)
+						continue
+					} else if err != nil {
+						return err
+					}
+					validationSecrets = append(validationSecrets, *secret)
+				}
+
+				verified := false
+				for _, pubSecret := range validationSecrets {
+					for k, data := range pubSecret.Data {
+						// search for public keys in the secret
+						if strings.HasSuffix(k, ".pub") {
+							verifier, err := soci.New(soci.WithPublicKey(data), authnKeychain)
+							if err != nil {
+								return err
+							}
+
+							signatures, _, err := verifier.VerifyImageSignatures(ctx, ref)
+							if err != nil {
+								continue
+							}
+
+							if signatures != nil {
+								verified = true
+								break
+							}
+						}
+					}
+				}
+
+				if !verified {
+					return fmt.Errorf("no matching signatures were found for the image %s", ref)
+				}
+
+				return nil
+			} else {
```

Comment:
Same note on returning early here, which it looks like you are already doing 🙂 (i.e. eliminate the `else`)